### PR TITLE
Visualisations and Utilities Introduced

### PR DESCRIPTION
**Pull request by Alex-Ian-Hamilton**
The primary goal of this PR is to introduce the utilities (general tools for use in the package) and the visualisations functions for viewing the extrapolation results in MayaVi.

### CODE REVIEW

File: solarextrapolation/classes.py [hunk 1]
```diff
@@ -13,9 +13,17 @@
 import pickle
 import time
 from datetime import datetime
+from astropy import units as u
+#from sunpy.sun._constants import physical_constants as con
+
+from sunpy.sun import constants
+from sunpy.sun import sun
```
Comment (Cadair) [inline solarextrapolation/classes.py]: style: you could write this as `from sunpy.sun import sun, constants`

File: solarextrapolation/classes.py [hunk 2]
```diff
@@ -13,9 +13,17 @@
...
+from sunpy.time import parse_time, is_time
+import warnings
+import inspect
 
 from astropy.table import Table
-#import astropy.units as u
+import astropy.units as u
```
Comment (Cadair) [inline solarextrapolation/classes.py]: duplicate units import.

File: solarextrapolation/classes.py [hunk 3]
```diff
@@ -195,31 +301,64 @@ def is_scalar(self):
...
-        xmin, xmax = -20.0, +20.0
-        return [xmin, xmax]
+    def units(self):
+        """
+        Image coordinate units along the x, y and z axes (cunit1/2/3).
+        """
 
+        # Define a triple, a named tuple object for returning values
+        from collections import namedtuple
+        Triple = namedtuple('Pair', 'x y z')
```
Comment (Cadair) [inline solarextrapolation/classes.py]: The string should read `'Triple'`

File: solarextrapolation/classes.py [hunk 4]
```diff
@@ -195,31 +301,64 @@ def is_scalar(self):
...
-        
-        xmin, xmax = -20.0, +20.0
-        return [xmin, xmax]
+    def units(self):
+        """
+        Image coordinate units along the x, y and z axes (cunit1/2/3).
+        """
 
+        # Define a triple, a named tuple object for returning values
+        from collections import namedtuple
```
Comment (Cadair) [inline solarextrapolation/classes.py]: this import should be at the top of the file.

File: solarextrapolation/utilities.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+from astropy import units as u
+import numpy as np
+import sunpy.map as mp
+from copy import deepcopy
+
+def decompose_ang_len(qua_input, **kwargs):
+    """
+    Function to help decompose quantities that have an equivilence between angles
+    and length, such as photospheric observational angles and object sizes.
+    The function uses an equivilence to cobnvert into eithe length or angle
```
Comment (Cadair) [inline solarextrapolation/utilities.py]: typo "convert"

Comment (Cadair) [inline solarextrapolation/utilities.py]: and "either"

File: solarextrapolation/utilities.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    physical_type units.
+
+    Parameters
+    ----------
+
```
Comment (Cadair) [inline solarextrapolation/utilities.py]: Parameters should also include a description of the arguments, not just kwargs. Also unless you have a reason for doing `**kwargs` you should just specify them in the function signature, as it makes it easier to inspect.

File: solarextrapolation/utilities.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    ----------
+    
+    working_units : astropy.unit, optional
+        Unit that will be used for internal working and the returned quantity.
+    Ensure that it is of the correct physical type, angle or length.
+    
+    equivalencies : astropy equivilence,
+        Equivilence used to relate the length and angle units.
+    
+    """
```
Comment (Cadair) [inline solarextrapolation/utilities.py]: Also, holy crap this function!!

You should write a whole suite of tests for this as a priority.

File: solarextrapolation/utilities.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        #length_unit = length_unit.value # Kept in-case it causes bugs
+    # The quantity to return
+    quantity =  value * length_unit ** length_exponent * angle_unit ** angle_exponent
+    # Change to the working unit if not dimensionless
+    if quantity.unit.physical_type is not (u.m / u.m).decompose().physical_type:
+        quantity.to(working_units)
+    return quantity.decompose()
+    
+
+def si_this_map_OLD(map):
```
Comment (Cadair) [inline solarextrapolation/utilities.py]: if you don't need this anymore, delete it.

File: solarextrapolation/utilities.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    return mp.Map((data, meta))    
+
+
+def si_this_map(map):
+    """
+    Basic function to create a deep copy of a map but with all units in SI.
+    """
+    # Find out the value units and convert this and data to SI
+    units = 1.0 * u.Unit(map.meta['bunit']).to(u.Tesla) * u.Tesla
+    print units
```
Comment (Cadair) [inline solarextrapolation/utilities.py]: superfluous print?

**Comment by Cadair** [comment]
Firstly, We will need to do some git foo next time you and I are free at the same time.

Secondly, I would break the visualisation function up into sections that the user can combine however they want. i.e. one for the fieldline visualisation, one for the oversize boundary data, and one for the axes etc. Then try and make those functions return the relevant mayavi objects so that the user can change the properties of them.

File: solarextrapolation/utilities.py (hunk 7 above)
Comment (Alex-Ian-Hamilton) [inline solarextrapolation/utilities.py]: I have... 60 lines of examples that I match up to a spreadsheet that laboriously calculates each step in a conversion.
But I'll put them in the testing suite also.

File: solarextrapolation/utilities.py (hunk 8 above)
Comment (Alex-Ian-Hamilton) [inline solarextrapolation/utilities.py]: I was leaving that until I've make that code testing suite-worthy.

File: solarextrapolation/utilities.py (hunk 9 above)
Comment (Alex-Ian-Hamilton) [inline solarextrapolation/utilities.py]: OLD function, to be removed once I have fully tested the second implementation, that SI's the length units too. I think that second part will be necessary in order to create a suitable dataset for all cuboid grid vector fields.

File: solarextrapolation/utilities.py (hunk 8 above)
Comment (Cadair) [inline solarextrapolation/utilities.py]: fair enough

File: solarextrapolation/utilities.py (hunk 7 above)
Comment (Cadair) [inline solarextrapolation/utilities.py]: that is what a unit test suite is supposed to do ;)

**Pull request merged**